File: app/src/main/java/org/deiverbum/app/ui/fragments/BibliaDataFragment.java
package org.deiverbum.app.ui.fragments;

import static org.deiverbum.app.utils.Constants.PACIENCIA;
import static org.deiverbum.app.utils.Constants.SEPARADOR;
import static org.deiverbum.app.utils.Constants.VOICE_INI;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.method.LinkMovementMethod;
import android.util.TypedValue;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.NavigationUI;

import org.deiverbum.app.R;
import org.deiverbum.app.data.wrappers.DataWrapper;
import org.deiverbum.app.databinding.FragmentTextBinding;
import org.deiverbum.app.model.BibleBooks;
import org.deiverbum.app.utils.TextToSpeechCallback;
import org.deiverbum.app.utils.TtsManager;
import org.deiverbum.app.utils.Utils;
import org.deiverbum.app.utils.ZoomTextView;
import org.deiverbum.app.viewmodel.BibliaViewModel;

import java.util.Locale;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class BibliaDataFragment extends Fragment implements TextToSpeechCallback {
    private BibliaViewModel mViewModel;
    private FragmentTextBinding binding;
    private ZoomTextView mTextView;
    private ProgressBar progressBar;
    private boolean isVoiceOn;
    private StringBuilder sbReader;
    private SeekBar seekBar;
    private boolean isReading = false;
    private Menu audioMenu;
    private MenuItem voiceItem;
    public static ActionMode mActionMode;
    private TtsManager mTtsManager;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        requireActivity().addMenuProvider(new MenuProvider() {
            @Override
            public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
                menuInflater.inflate(R.menu.toolbar_menu, menu);
                voiceItem=menu.findItem(R.id.item_voz);
                voiceItem.setVisible(isVoiceOn);
                if (isReading) {
                    voiceItem.setVisible(false);
                }
            }

            @Override
            public boolean onMenuItemSelected(@NonNull MenuItem item) {
                if (item.getItemId() == R.id.item_voz) {
                    if (mActionMode == null) {
                        mActionMode =
                                requireActivity().startActionMode(mActionModeCallback);
                    }
                    readText();
                    isReading = true;
                    voiceItem.setVisible(false);
                    //item.setVisible(!isReading);
                    requireActivity().invalidateOptionsMenu();
                    return true;
                }
                NavController navController = NavHostFragment.findNavController(requireParentFragment());
                return NavigationUI.onNavDestinationSelected(item, navController);
            }
        }, getViewLifecycleOwner(), Lifecycle.State.RESUMED);
        binding = FragmentTextBinding.inflate(inflater, container, false);
        View root = binding.getRoot();
        setConfiguration();
        observeData();
        return root;
    }

    private void setConfiguration() {
        mViewModel =
                new ViewModelProvider(this).get(BibliaViewModel.class);
        mTextView = binding.include.tvZoomable;
        mTextView.setMovementMethod(LinkMovementMethod.getInstance());
        progressBar = binding.progressBar;
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getActivity());
        float fontSize = Float.parseFloat(prefs.getString("font_size", "18"));
        mTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, fontSize);
        String fontFamily = String.format(new Locale("es"),"fonts/%s",prefs.getString("font_name", "robotoslab_regular.ttf"));
        Typeface tf= Typeface.createFromAsset(requireActivity().getAssets(),fontFamily);
        mTextView .setTypeface(tf);
        isVoiceOn = prefs.getBoolean("voice", true);
    }

    private void observeData() {
        int param = requireArguments().getInt("bookId");
        mTextView.setText(PACIENCIA);
        mViewModel.getLibro(param).observe(getViewLifecycleOwner(), data -> {
            progressBar.setVisibility(View.GONE);
            if (data.status == DataWrapper.Status.SUCCESS) {
                BibleBooks libro=data.getData();
                mTextView.setText(libro.getForView());
                if (isVoiceOn) {
                    sbReader = new StringBuilder(VOICE_INI);
                    sbReader.append(libro.getForRead());
                    //setPlayerButton();
                }
            } else {
                mTextView.setText(Utils.fromHtml(data.getError()));
            }
        });
    }

    private String prepareForRead() {
        String notQuotes = Utils.stripQuotation(sbReader.toString());
        return String.valueOf(Utils.fromHtml(notQuotes));
    }

    @Override
    public void onCompleted() {
    }

    @Override
    public void onError() {
    }

    private final ActionMode.Callback mActionModeCallback = new ActionMode.Callback() {
        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            int menuItem = item.getItemId();
            if (menuItem == R.id.audio_play) {
                readText();
                audioMenu.findItem(R.id.audio_pause).setVisible(true);
                audioMenu.findItem(R.id.audio_stop).setVisible(true);
                item.setVisible(false);
                return true;
            }
            if (menuItem == R.id.audio_pause) {
                mTtsManager.pause();
                audioMenu.findItem(R.id.audio_resume).setVisible(true);
                item.setVisible(false);
                return true;
            }
            if (menuItem == R.id.audio_resume) {
                mTtsManager.resume();
                audioMenu.findItem(R.id.audio_pause).setVisible(true);
                item.setVisible(false);
                return true;
            }
            if (menuItem == R.id.audio_stop) {
                mTtsManager.stop();
                audioMenu.findItem(R.id.audio_play).setVisible(true);
                audioMenu.findItem(R.id.audio_pause).setVisible(false);
                audioMenu.findItem(R.id.audio_resume).setVisible(false);
                item.setVisible(false);
                return true;
            }
            return false;
        }

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.contextual_action_bar, menu);
            audioMenu = menu;
            @SuppressLint("InflateParams")
            View view = LayoutInflater.from(getContext()).inflate(R.layout.seekbar, null);
            mode.setCustomView(view);
            seekBar = view.findViewById(R.id.seekbar);
            return true;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            mActionMode = null;
            cleanTTS();
            setPlayerButton();
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }
    };

    private void setPlayerButton() {
        voiceItem.setVisible(isVoiceOn);
    }

    private void readText() {
        mTtsManager = new TtsManager(getContext(), prepareForRead(), SEPARADOR, (current, max) -> {
            seekBar.setProgress(current);
            seekBar.setMax(max);
        });

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (mTtsManager == null) return;
                mTtsManager.changeProgress(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
        mTtsManager.start();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mActionMode != null) {
            mActionMode.finish();
        }
        cleanTTS();
        binding = null;
    }

    private void cleanTTS() {
        if (mTtsManager != null) {
            mTtsManager.close();
        }
    }
}

File: app/src/main/java/org/deiverbum/app/core/presentation/biblia/BibliaDataFragment.kt
package org.deiverbum.app.core.presentation.biblia

import android.annotation.SuppressLint
import android.graphics.Typeface
import android.util.TypedValue
import android.view.*
import android.widget.ProgressBar
import android.widget.SeekBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.NavigationUI
import androidx.preference.PreferenceManager
import androidx.viewbinding.ViewBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.deiverbum.app.R
import org.deiverbum.app.core.model.BibleBookRequest
import org.deiverbum.app.core.presentation.base.BaseFragment
import org.deiverbum.app.databinding.FragmentTextBinding
import org.deiverbum.app.util.Constants
import org.deiverbum.app.util.TtsManager
import org.deiverbum.app.util.ZoomTextView
import java.util.*


/**
 * Este Fragmento coordina la obtención de los datos del módulo **Biblia**.
 *
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1.3
 */

@AndroidEntryPoint
class BibliaDataFragment : BaseFragment<FragmentTextBinding>() {
    private val mViewModel: BibliaViewModel by viewModels()
    private lateinit var mTextView: ZoomTextView
    private var progressBar: ProgressBar? = null
    private var seekBar: SeekBar? = null
    private var isReading = false
    private var isVoiceOn = false
    private var hasInvitatory = false
    private var sbReader: StringBuilder? = null
    private var audioMenu: Menu? = null
    private var voiceItem: MenuItem? = null
    private var mTtsManager: TtsManager? = null
    private var mainMenu: Menu? = null
    private lateinit var mRequest: BibleBookRequest
    private var mActionMode: ActionMode?=null

    override fun constructViewBinding(): ViewBinding = FragmentTextBinding.inflate(layoutInflater)

    override fun init(viewBinding: ViewBinding) {
        setConfiguration()
        setMenu()
        mViewModel.loadData(mRequest)
        fetchData()
    }

    private fun setConfiguration() {
        updateSubTitle()
        val hourId = requireArguments().getInt("bookId")
        mTextView = getViewBinding().include.tvZoomable
        progressBar = getViewBinding().progressBar
        val sp = PreferenceManager.getDefaultSharedPreferences(requireActivity().applicationContext)
        val fontSize = sp?.getString("font_size", "18")!!.toFloat()
        val fontFamily = String.format(
            Locale("es"),
            "fonts/%s",
            sp.getString("font_name", "robotoslab_regular.ttf")
        )
        val tf = Typeface.createFromAsset(requireActivity().assets, fontFamily)
        mTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, fontSize)
        mTextView.typeface = tf
        hasInvitatory = sp.getBoolean("invitatorio", false)
        isVoiceOn = sp.getBoolean("voice", true)
        mRequest = BibleBookRequest(hourId)
    }

    private fun fetchData() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                mViewModel.uiState.collect { state ->
                    when (state) {
                        is BibliaViewModel.BibliaUiState.Loaded -> onLoaded(state.itemState)
                        is BibliaViewModel.BibliaUiState.Error -> showError(state.message)
                        else -> showLoading()
                    }
                }
            }
        }
    }

    private fun onLoaded(todayItemUiState: BibliaItemUiState) {
        todayItemUiState.run {
            getViewBinding().progressBar.visibility = View.GONE
            mTextView.text = bibleBookResponse.getForView(isNightMode())
        }
    }

    private fun showLoading() {
        mTextView.text = Constants.PACIENCIA
    }

    private fun showError(stringRes: String) {
        mTextView.text = stringRes
        Toast.makeText(requireContext(), stringRes, Toast.LENGTH_SHORT).show()
    }

    val mActionModeCallback: ActionMode.Callback = object : ActionMode.Callback {
        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            val menuItem = item.itemId
            if (menuItem == R.id.audio_play) {
                readText()
                audioMenu?.findItem(R.id.audio_pause)?.isVisible = true
                audioMenu?.findItem(R.id.audio_stop)?.isVisible = true
                item.isVisible = false
                return true
            }
            if (menuItem == R.id.audio_pause) {
                mTtsManager?.pause()
                audioMenu?.findItem(R.id.audio_resume)?.isVisible = true
                item.isVisible = false
                return true
            }
            if (menuItem == R.id.audio_resume) {
                mTtsManager?.resume()
                audioMenu?.findItem(R.id.audio_pause)?.isVisible = true
                item.isVisible = false
                return true
            }
            if (menuItem == R.id.audio_stop) {
                mTtsManager?.stop()
                audioMenu?.findItem(R.id.audio_play)?.isVisible = true
                audioMenu?.findItem(R.id.audio_pause)?.isVisible = false
                audioMenu?.findItem(R.id.audio_resume)?.isVisible = false
                item.isVisible = false
                return true
            }
            return false
        }

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.contextual_action_bar, menu)
            audioMenu = menu
            @SuppressLint("InflateParams") val view: View =
                LayoutInflater.from(context).inflate(R.layout.seekbar, null)
            mode.customView = view
            seekBar = view.findViewById(R.id.seekbar)
            return true
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            mActionMode = null
            cleanTTS()
            setPlayerButton()
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            return false
        }
    }


    private fun setMenu() {
        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.toolbar_menu, menu)
                mainMenu = menu
                voiceItem = menu.findItem(R.id.item_voz)
                voiceItem!!.isVisible = isVoiceOn
                if (isReading) {
                    voiceItem!!.isVisible = false
                }
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                return when (menuItem.itemId) {
                    android.R.id.home -> {
                        val navController =
                            NavHostFragment.findNavController(requireParentFragment())
                        navController.popBackStack()
                        true
                    }
                    R.id.item_voz -> {
                        if (mActionMode == null) {
                            mActionMode =
                                requireActivity().startActionMode(mActionModeCallback)
                        }
                        readText()
                        isReading = true
                        voiceItem?.isVisible = false
                        requireActivity().invalidateOptionsMenu()
                        true
                    }
                    else -> {
                        val navController =
                            NavHostFragment.findNavController(requireParentFragment())
                        NavigationUI.onNavDestinationSelected(menuItem, navController)
                    }
                }

            }
        }, viewLifecycleOwner)
    }

    private fun setPlayerButton() {
        voiceItem!!.isVisible = isVoiceOn
    }

    private fun readText() {
        mTtsManager = TtsManager(
            context,
            sbReader.toString(),
            Constants.SEPARADOR
        ) { current: Int, max: Int ->
            seekBar!!.progress = current
            seekBar!!.max = max
        }
        seekBar!!.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (mTtsManager == null) return
                mTtsManager!!.changeProgress(progress)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        mTtsManager!!.start()
    }

    fun onError() {}

    private fun cleanTTS() {
        mTtsManager?.close()
    }

    private fun updateSubTitle() {
        val actionBar = (requireActivity() as AppCompatActivity).supportActionBar
        actionBar?.subtitle = ""
    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (mActionMode != null) {
            mActionMode!!.finish()
        }
        cleanTTS()
    }

}